### PR TITLE
mkdir: fixed not respecting set umask

### LINKED ISSUE #3119: mkdir: does not respect umask
GNU mkdir:
```
$ umask 000
$ mkdir a
$ ls -ld a
drwxrwxrwx 2 jeffrey jeffrey 4096 Feb 11 22:11 a
```

uutils mkdir:
```
$ umask 000
$ ./target/debug/mkdir a
$ ls -ld a
drwxr-xr-x 2 jeffrey jeffrey 4096 Feb 11 22:11 a
```

### PR DESCRIPTION
**Pull request by pyoky**
Resolves #3119.

Removed the permissions mode argument's hardcoded `default_value()` of `755`, instead using `uutils::mode::get_umask()` when argument isn't specified.

Before:
```
$ umask 000 && ./coreutils mkdir new_dir && ls -l | grep new_dir
drwxr-xr-x  2 pyoky pyoky     4096 Feb 17 08:30 new_dir
```

After:
```
$ umask 000 && ./coreutils mkdir new_dir && ls -l | grep new_dir
d---------  2 pyoky pyoky     4096 Feb 17 08:29 new_dir
```
This mirrors the behavior of GNU `mkdir`.

### CODE REVIEW

**Comment by sylvestre** [comment]
Looks good but could you please add a test?

**Comment by pyoky** [comment]
@sylvestre I'm having issues writing a test that would modify the `umask`, create a new directory, and test if the permissions are expected. `umask` is a shell builtin, meaning that `TestScenario` can't execute it as a command. This is what I'm trying instead:
```
#[test]
#[cfg(not(windows))]
fn test_umask_compliance() {
    let scene = TestScenario::new(util_name!());
    let at = scene.fixtures.clone();
    let umask_out = scene.cmd("sh")
         .arg("-c")
         .arg("umask")
         .arg("000")
         .succeeds();
    let mkdir_out = scene.ucmd().arg(TEST_DIR8).succeeds();
    let perms = at.metadata(TEST_DIR8).permissions().mode();
    assert_eq!(perms, 0o40000);
}
```
but this has the issue that `umask` is dependent on the shell, which means as `sh` exits the modified `umask` disappears along with it.

I can write one that checks if it respects the already set  `umask` but it won't be effective. What do you suggest?

**Comment by tertsdiepraam** [comment]
I think that's a good solution. On my machine, this seems to give the correct umask:
```
❯ sh -c "umask 000 && mkdir a"
❯ mkdir b
❯ ls -l
total 0
drwxrwxrwx 2 terts terts 40 19 feb 01:48 a
drwxr-xr-x 2 terts terts 40 19 feb 01:48 b
```

**Comment by sylvestre** [comment]
This fails:
```
error[E0308]: mismatched types
   --> /Users/runner/work/coreutils/coreutils/tests/by-util/test_mkdir.rs:113:45
    |
113 |         let original_umask = unsafe { umask(umask_set) };
    |                                             ^^^^^^^^^ expected `u16`, found `u32`
    |
help: you can convert a `u32` to a `u16` and panic if the converted value doesn't fit
    |
113 |         let original_umask = unsafe { umask(umask_set.try_into().unwrap()) };
    |                                                      ++++++++++++++++++++
```
could you please have a look? thanks

**Comment by pyoky** [comment]
I changed the `umask_set` to `mode_t`. Test should work now.

**Comment by sylvestre** [comment]
A few tests are failing, could you please have a look? Thanks

**Comment by sylvestre** [comment]
ping?

**Comment by pyoky** [comment]
@sylvestre Yup I'm working on it. I'll push a commit in a day or two.

**Comment by pyoky** [comment]
@tertsdiepraam It seems like my use of `libc::umask()` messes with permissions of test directories with messes up permissions for tests that come after. The test now only test umasks with a reasonable number of permission combinations. (Testing all `0o0`~`0o777` was unnecessary anyways.)

While testing I also found a slight bug in `test_pwd.rs` where the directory name in a test command was not escaped properly. It was a minor bug so I included in the commit with descriptions.

Lints should also be fixed. Sorry for my mistakes - I'm still new to contributions.

**Comment by tertsdiepraam** [comment]
No worries! Mistakes are basically expected :) Great work on the fix!

File: tests/by-util/test_mkdir.rs [hunk 1]
```diff
@@ -13,6 +17,7 @@ static TEST_DIR8: &str = "mkdir_test8/mkdir_test8_1/mkdir_test8_2";
 static TEST_DIR9: &str = "mkdir_test9/../mkdir_test9_1/../mkdir_test9_2";
 static TEST_DIR10: &str = "mkdir_test10/.";
 static TEST_DIR11: &str = "mkdir_test11/..";
+static TEST_DIR12: &str = "mkdir_test12";
```
Comment (tertsdiepraam) [inline tests/by-util/test_mkdir.rs:20]:
```suggestion
#[cfg(not(windows))]
static TEST_DIR12: &str = "mkdir_test12";
```

**Review by tertsdiepraam — COMMENTED** [review]
I think this should fix the last clippy error

**Comment by pyoky** [comment]
@tertsdiepraam Is there an issue still with checks? I'm seeing the tests are canceled. If there's still issues with merging let me know.

**Comment by tertsdiepraam** [comment]
I'm rerunning those checks. If they pass, I think this can get merged.

File: tests/by-util/test_mkdir.rs [hunk 2]
```diff
@@ -151,3 +157,26 @@ fn test_mkdir_trailing_dot() {
...
+        let perms = at.metadata(TEST_DIR12).permissions().mode() as mode_t;
+
+        assert_eq!(perms, (!umask_set & 0o0777) + 0o40000); // before compare, add the set GUID, UID bits
+        unsafe {
+            umask(original_umask);
+        } // set umask back to original
+    }
+
+    for i in 0o0..0o027 {
+        // tests all permission combinations
```
Comment (jfinkels) [inline tests/by-util/test_mkdir.rs:179]: These are not all the possible permission settings. That's fine, let's just make sure the comment matches the code.

File: tests/by-util/test_pwd.rs [hunk 3]
```diff
@@ -26,7 +26,7 @@ fn test_deleted_dir() {
     let output = Command::new("sh")
         .arg("-c")
         .arg(format!(
-            "cd '{}'; mkdir foo; cd foo; rmdir ../foo; exec {} {}",
+            "cd '{}'; mkdir foo; cd foo; rmdir ../foo; exec '{}' {}",
```
Comment (jfinkels) [inline tests/by-util/test_pwd.rs:29]: Is this change related to the changes in `mkdir`?

**Review by jfinkels — APPROVED** [review]
Cool

File: tests/by-util/test_pwd.rs (hunk 3 above)
Comment (pyoky) [reply tests/by-util/test_pwd.rs:29]: The non-escaped pathnames caused problems in my machine when testing.

**Pull request merged**